frontend/public/components/app.jsx

```diff
@@ -132,13 +132,57 @@ const App = (props) => {
     }
   }, []);
 
+  const onCSPViolation = React.useCallback((event) => {
+    // eslint-disable-next-line no-console
+    console.warn('Content-Security-Policy: violation detected', event);
+
+    // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent
+    const cspReportObject = _.pick(event, [
+      // The URI of the resource that was blocked because it violates a policy.
+      'blockedURI',
+      // The column number in the document or worker at which the violation occurred.
+      'columnNumber',
+      // Whether the user agent is configured to enforce or just report the policy violation.
+      'disposition',
+      // The URI of the document or worker in which the violation occurred.
+      'documentURI',
+      // The directive that was violated.
+      'effectiveDirective',
+      // The line number in the document or worker at which the violation occurred.
+      'lineNumber',
+      // The policy whose enforcement caused the violation.
+      'originalPolicy',
+      // The URL for the referrer of the resources whose policy was violated, or null.
+      'referrer',
+      // A sample of the resource that caused the violation, usually the first 40 characters.
+      // This will only be populated if the resource is an inline script, event handler or style.
+      'sample',
+      // If the violation occurred as a result of a script, this will be the URL of the script.
+      'sourceFile',
+      // HTTP status code of the document or worker in which the violation occurred.
+      'statusCode',
+    ]);
+
+    const fireTelemetryEvent = useTelemetry();
+
+    // TODO(vojtech): attempt to infer Console plugin name from CSP report object
+    fireTelemetryEvent('Content-Security-Policy violation', cspReportObject);
```

Comment:
Is the violation automatically logged to the JS console as well?

We should discuss if there is a way during development mode to make it really obvious a violation occurred without the dev needing to explicitly check the console. Also we should think about the plugin development use case. We want to make sure plugin developers know they have violations that will break their plugin when we turn on enforcement.

Reply:
good idea. Maybe having a similar warnings like we do have with admission webhook, but just in the DEV mode?